File: src/main/java/me/alegian/thavma/impl/init/registries/deferred/T7CreativeModeTabs.java
package me.alegian.thavma.impl.init.registries.deferred;

import me.alegian.thavma.impl.Thavma;
import net.minecraft.core.registries.Registries;
import net.minecraft.network.chat.Component;
import net.minecraft.world.item.CreativeModeTab;
import net.neoforged.neoforge.registries.DeferredRegister;

public class T7CreativeModeTabs {
  public static final DeferredRegister<CreativeModeTab> REGISTRAR = DeferredRegister.create(Registries.CREATIVE_MODE_TAB, Thavma.MODID);

  static {
    T7CreativeModeTabs.REGISTRAR.register(
        Thavma.MODID,
        () -> CreativeModeTab
            .builder()
            .title(Component.translatable(Thavma.MODID))
            .icon(() -> T7Items.THAUMONOMICON.get().getDefaultInstance())
            .displayItems((parameters, output) -> {
              output.accept(T7Blocks.INSTANCE.getAURA_NODE().get());
              output.accept(T7Blocks.INSTANCE.getESSENTIA_CONTAINER().get());

              output.accept(T7Blocks.INSTANCE.getCRUCIBLE().get());
              output.accept(T7Blocks.INSTANCE.getARCANE_WORKBENCH().get());
              output.accept(T7Blocks.INSTANCE.getMATRIX().get());
              output.accept(T7Blocks.INSTANCE.getPILLAR().get());
              output.accept(T7Blocks.INSTANCE.getPEDESTAL().get());
              output.accept(T7Blocks.INSTANCE.getPEDESTAL().get());

              output.accept(T7Blocks.INSTANCE.getELEMENTAL_STONE().get());

              for (var infusedBlock : T7Blocks.INSTANCE.getINFUSED_BLOCKS())
                output.accept(infusedBlock.get());

              output.accept(T7Blocks.INSTANCE.getARCANUM_BLOCK().get());
              output.accept(T7Blocks.INSTANCE.getORICHALCUM_BLOCK().get());

              output.accept(T7Blocks.INSTANCE.getGREATWOOD_LEAVES().get());
              output.accept(T7Blocks.INSTANCE.getGREATWOOD_LOG().get());
              output.accept(T7Blocks.INSTANCE.getGREATWOOD_PLANKS().get());
              output.accept(T7Blocks.INSTANCE.getGREATWOOD_SAPLING().get());

              output.accept(T7Blocks.INSTANCE.getSILVERWOOD_LEAVES().get());
              output.accept(T7Blocks.INSTANCE.getSILVERWOOD_SAPLING().get());
              output.accept(T7Blocks.INSTANCE.getSILVERWOOD_LOG().get());
              output.accept(T7Blocks.INSTANCE.getSILVERWOOD_PLANKS().get());

              output.accept(T7Items.IRON_HANDLE.get());
              output.accept(T7Items.GOLD_HANDLE.get());
              output.accept(T7Items.ORICHALCUM_HANDLE.get());
              output.accept(T7Items.ARCANUM_HANDLE.get());

              output.accept(T7Items.EYE_OF_WARDEN.get());

              output.accept(T7Items.GREATWOOD_CORE.get());
              output.accept(T7Items.SILVERWOOD_CORE.get());

              output.accept(T7Items.RUNE.get());
              output.accept(T7Items.ARCANUM_INGOT.get());
              output.accept(T7Items.ARCANUM_NUGGET.get());
              output.accept(T7Items.ORICHALCUM_INGOT.get());
              output.accept(T7Items.ORICHALCUM_NUGGET.get());
              output.accept(T7Items.RESEARCH_SCROLL.get());
              output.accept(T7Items.COMPLETED_RESEARCH.get());
              output.accept(T7Items.OCULUS.get());
              output.accept(T7Items.THAUMONOMICON.get());

              output.accept(T7Items.wandOrThrow(WandHandleMaterials.INSTANCE.getIRON().get(), WandCoreMaterials.INSTANCE.getWOOD().get()));
              output.accept(T7Items.wandOrThrow(WandHandleMaterials.INSTANCE.getGOLD().get(), WandCoreMaterials.INSTANCE.getGREATWOOD().get()));
              output.accept(T7Items.wandOrThrow(WandHandleMaterials.INSTANCE.getARCANUM().get(), WandCoreMaterials.INSTANCE.getSILVERWOOD().get()));

              output.accept(T7Items.GOGGLES.get());
              output.accept(T7Items.GOGGLES_ACCESSORY.get());
              output.accept(T7Items.DAWN_CHARM.get());
              output.accept(T7Items.RESEARCHER_CHESTPLATE.get());
              output.accept(T7Items.RESEARCHER_LEGGINGS.get());
              output.accept(T7Items.RESEARCHER_BOOTS.get());

              output.accept(T7Items.ARCANUM_HELMET.get());
              output.accept(T7Items.ARCANUM_CHESTPLATE.get());
              output.accept(T7Items.ARCANUM_LEGGINGS.get());
              output.accept(T7Items.ARCANUM_BOOTS.get());

              output.accept(T7Items.CUSTOS_ARCANUM_HELMET.get());
              output.accept(T7Items.CUSTOS_ARCANUM_CHESTPLATE.get());
              output.accept(T7Items.CUSTOS_ARCANUM_LEGGINGS.get());
              output.accept(T7Items.CUSTOS_ARCANUM_BOOTS.get());

              output.accept(T7Items.IGNIS_TESTA.get());
              output.accept(T7Items.AER_TESTA.get());
              output.accept(T7Items.TERRA_TESTA.get());
              output.accept(T7Items.AQUA_TESTA.get());
              output.accept(T7Items.ORDO_TESTA.get());
              output.accept(T7Items.PERDITIO_TESTA.get());

              output.accept(T7Items.ARCANUM_AXE.get());
              output.accept(T7Items.ARCANUM_PICKAXE.get());
              output.accept(T7Items.ARCANUM_HAMMER.get());
              output.accept(T7Items.ARCANUM_SHOVEL.get());
              output.accept(T7Items.ARCANUM_HOE.get());

              output.accept(T7Items.ARCANUM_SWORD.get());
              output.accept(T7Items.ARCANUM_KATANA.get());
              output.accept(T7Items.ZEPHYR.get());
            })
            .build()
    );
  }
}

File: src/main/java/me/alegian/thavma/impl/init/registries/deferred/T7CreativeModeTabs.kt
package me.alegian.thavma.impl.init.registries.deferred

import me.alegian.thavma.impl.Thavma
import me.alegian.thavma.impl.init.registries.deferred.T7Blocks.ARCANE_WORKBENCH
import me.alegian.thavma.impl.init.registries.deferred.T7Blocks.ARCANUM_BLOCK
import me.alegian.thavma.impl.init.registries.deferred.T7Blocks.AURA_NODE
import me.alegian.thavma.impl.init.registries.deferred.T7Blocks.CRUCIBLE
import me.alegian.thavma.impl.init.registries.deferred.T7Blocks.ELEMENTAL_STONE
import me.alegian.thavma.impl.init.registries.deferred.T7Blocks.ESSENTIA_CONTAINER
import me.alegian.thavma.impl.init.registries.deferred.T7Blocks.GREATWOOD_LEAVES
import me.alegian.thavma.impl.init.registries.deferred.T7Blocks.GREATWOOD_LOG
import me.alegian.thavma.impl.init.registries.deferred.T7Blocks.GREATWOOD_PLANKS
import me.alegian.thavma.impl.init.registries.deferred.T7Blocks.GREATWOOD_SAPLING
import me.alegian.thavma.impl.init.registries.deferred.T7Blocks.INFUSED_DEEPSLATES
import me.alegian.thavma.impl.init.registries.deferred.T7Blocks.INFUSED_STONES
import me.alegian.thavma.impl.init.registries.deferred.T7Blocks.MATRIX
import me.alegian.thavma.impl.init.registries.deferred.T7Blocks.ORICHALCUM_BLOCK
import me.alegian.thavma.impl.init.registries.deferred.T7Blocks.PEDESTAL
import me.alegian.thavma.impl.init.registries.deferred.T7Blocks.PILLAR
import me.alegian.thavma.impl.init.registries.deferred.T7Blocks.SILVERWOOD_LEAVES
import me.alegian.thavma.impl.init.registries.deferred.T7Blocks.SILVERWOOD_LOG
import me.alegian.thavma.impl.init.registries.deferred.T7Blocks.SILVERWOOD_PLANKS
import me.alegian.thavma.impl.init.registries.deferred.T7Blocks.SILVERWOOD_SAPLING
import me.alegian.thavma.impl.init.registries.deferred.WandCoreMaterials.GREATWOOD
import me.alegian.thavma.impl.init.registries.deferred.WandCoreMaterials.SILVERWOOD
import me.alegian.thavma.impl.init.registries.deferred.WandCoreMaterials.WOOD
import me.alegian.thavma.impl.init.registries.deferred.WandHandleMaterials.ARCANUM
import me.alegian.thavma.impl.init.registries.deferred.WandHandleMaterials.GOLD
import me.alegian.thavma.impl.init.registries.deferred.WandHandleMaterials.IRON
import net.minecraft.core.registries.Registries
import net.minecraft.network.chat.Component
import net.minecraft.world.item.CreativeModeTab
import net.neoforged.neoforge.registries.DeferredRegister

object T7CreativeModeTabs {
    val REGISTRAR = DeferredRegister.create(Registries.CREATIVE_MODE_TAB, Thavma.MODID)

    init {
        REGISTRAR.register(
            Thavma.MODID
        ) { ->
            CreativeModeTab
                .builder()
                .title(Component.translatable(Thavma.MODID))
                .icon { T7Items.THAUMONOMICON.get().defaultInstance }
                .displayItems { _, output ->
                    output.accept(AURA_NODE.get())
                    output.accept(ESSENTIA_CONTAINER.get())

                    output.accept(CRUCIBLE.get())
                    output.accept(ARCANE_WORKBENCH.get())
                    output.accept(MATRIX.get())
                    output.accept(PILLAR.get())
                    output.accept(PEDESTAL.get())
                    output.accept(PEDESTAL.get())

                    output.accept(ELEMENTAL_STONE.get())

                    for(infusedStone in INFUSED_STONES.values) output.accept(infusedStone.get())
                    for(infusedStone in INFUSED_DEEPSLATES.values) output.accept(infusedStone.get())

                    output.accept(ARCANUM_BLOCK.get())
                    output.accept(ORICHALCUM_BLOCK.get())

                    output.accept(GREATWOOD_LEAVES.get())
                    output.accept(GREATWOOD_LOG.get())
                    output.accept(GREATWOOD_PLANKS.get())
                    output.accept(GREATWOOD_SAPLING.get())

                    output.accept(SILVERWOOD_LEAVES.get())
                    output.accept(SILVERWOOD_SAPLING.get())
                    output.accept(SILVERWOOD_LOG.get())
                    output.accept(SILVERWOOD_PLANKS.get())

                    output.accept(T7Items.IRON_HANDLE)
                    output.accept(T7Items.GOLD_HANDLE)
                    output.accept(T7Items.ORICHALCUM_HANDLE)
                    output.accept(T7Items.ARCANUM_HANDLE)

                    output.accept(T7Items.EYE_OF_WARDEN)

                    output.accept(T7Items.GREATWOOD_CORE)
                    output.accept(T7Items.SILVERWOOD_CORE)

                    output.accept(T7Items.RUNE)
                    output.accept(T7Items.ARCANUM_INGOT)
                    output.accept(T7Items.ARCANUM_NUGGET)
                    output.accept(T7Items.ORICHALCUM_INGOT)
                    output.accept(T7Items.ORICHALCUM_NUGGET)
                    output.accept(T7Items.RESEARCH_SCROLL)
                    output.accept(T7Items.COMPLETED_RESEARCH)
                    output.accept(T7Items.OCULUS)
                    output.accept(T7Items.THAUMONOMICON)

                    output.accept(T7Items.wandOrThrow(IRON.get(), WOOD.get()))
                    output.accept(
                        T7Items.wandOrThrow(
                            GOLD.get(),
                            GREATWOOD.get()
                        )
                    )
                    output.accept(
                        T7Items.wandOrThrow(
                            ARCANUM.get(),
                            SILVERWOOD.get()
                        )
                    )

                    output.accept(T7Items.GOGGLES)
                    output.accept(T7Items.GOGGLES_ACCESSORY)
                    output.accept(T7Items.DAWN_CHARM)
                    output.accept(T7Items.RESEARCHER_CHESTPLATE)
                    output.accept(T7Items.RESEARCHER_LEGGINGS)
                    output.accept(T7Items.RESEARCHER_BOOTS)

                    output.accept(T7Items.ARCANUM_HELMET)
                    output.accept(T7Items.ARCANUM_CHESTPLATE)
                    output.accept(T7Items.ARCANUM_LEGGINGS)
                    output.accept(T7Items.ARCANUM_BOOTS)

                    output.accept(T7Items.CUSTOS_ARCANUM_HELMET)
                    output.accept(T7Items.CUSTOS_ARCANUM_CHESTPLATE)
                    output.accept(T7Items.CUSTOS_ARCANUM_LEGGINGS)
                    output.accept(T7Items.CUSTOS_ARCANUM_BOOTS)

                    for(testa in T7Items.TESTAS.values) output.accept(testa)

                    output.accept(T7Items.ARCANUM_AXE)
                    output.accept(T7Items.ARCANUM_PICKAXE)
                    output.accept(T7Items.ARCANUM_HAMMER)
                    output.accept(T7Items.ARCANUM_SHOVEL)
                    output.accept(T7Items.ARCANUM_HOE)

                    output.accept(T7Items.ARCANUM_SWORD)
                    output.accept(T7Items.ARCANUM_KATANA)
                    output.accept(T7Items.ZEPHYR)

                    output.accept(T7Items.ANGRY_ZOMBIE_SPAWN_EGG)
                }
                .build()
        }
    }
}
